dronegen/aws.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import "path/filepath"
+
+// awsRoleSettings contains the information necessary to assume an AWS Role
+//
+// This is intended to be imbedded, please use the kubernetes/mac/windows versions
+// with their corresponding pipelines.
+type awsRoleSettings struct {
+	awsAccessKeyId     value
+	awsSecretAccessKey value
+	role               value
+}
+
+// kuberentesRoleSettings contains the info necessary to assume an AWS role and save the credentials to a volume that later steps can use
+type kubernetesRoleSettings struct {
+	awsRoleSettings
+	configVolume volumeRef
+}
+
+// macRoleSettings contains the info necessary to assume an AWS role and save the credentials to a path that later steps can use
+type macRoleSettings struct {
+	awsRoleSettings
+	configPath string
+}
+
+// kuberentesS3Settings contains all info needed to download from S3 in a kubernetes pipeline
+type kubernetesS3Settings struct {
+	region       string
+	source       string
+	target       string
+	configVolume volumeRef
+}
+
+// assumeRoleCommands is a helper to build the role assumtipn commands on a *nix platform
+func assumeRoleCommands(configPath string) []string {
+	assumeRoleCmd := `printf "[default]\naws_access_key_id = %s\naws_secret_access_key = %s\naws_session_token = %s" \
+  $(aws sts assume-role \
+    --role-arn "$AWS_ROLE" \
+    --role-session-name $(echo "drone-${DRONE_REPO}/${DRONE_BUILD_NUMBER}" | sed "s|/|-|g") \
```

Comment:
I assume `sed "s|/|-|g"` is supposed to clean the repo name, but it will replace the `/` between repo and build number as well?

Reply:
Yes -- that is intentional.  Role session names cannot have a `/` in them.  I can change it to have a `-` in the first place if you'd prefer.

https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html

> The regex used to validate this parameter is a string of characters consisting of upper- and lower-case alphanumeric characters with no spaces. You can also include underscores or any of the following characters: =,.@-
> 
> Type: String
> 
> Length Constraints: Minimum length of 2. Maximum length of 64.
> 
> Pattern: [\w+=,.@-]*

Reply:
updated to be consistent about using `-` in:
* https://github.com/gravitational/teleport/pull/17201/commits/285036df8f36247350a5ebd2c110d7eec7f73472
* https://github.com/gravitational/teleport/pull/17201/commits/0c15fa95cfaaa5c94925b8a8b4abaae268f812f8